api/src/websocket/controllers/hooks.ts

```diff
@@ -0,0 +1,130 @@
+import emitter from '../../emitter';
+import { getMessenger } from '../../messenger';
+import type { WebSocketEvent } from '../messages';
+
+let actionsRegistered = false;
+
+export function registerWebsocketEvents() {
+	if (actionsRegistered) return;
+	actionsRegistered = true;
+
+	registerActionHooks([
+		'items',
+		'activity',
+		'collections',
+		'folders',
+		'permissions',
+		'presets',
+		'revisions',
+		'roles',
+		'settings',
+		'users',
+		'webhooks',
+	]);
+	registerFieldsHooks();
+	registerFilesHooks();
+	registerRelationsHooks();
+}
+
+function registerActionHooks(modules: string[]) {
+	// register event hooks that can be handled in an uniform manner
+	for (const module of modules) {
+		registerAction(module + '.create', ({ key, collection, payload = {} }) => ({
+			collection,
+			action: 'create',
+			key,
+			payload,
+		}));
+		registerAction(module + '.update', ({ keys, collection, payload = {} }) => ({
+			collection,
+			action: 'update',
+			keys,
+			payload,
+		}));
+		registerAction(module + '.delete', ({ keys, collection, payload = [] }) => ({
+			collection,
+			action: 'delete',
+			keys,
+			payload,
+		}));
+	}
+}
+
+function registerFieldsHooks() {
+	// exception for field updates that don't report `directus_fields` as being the collection
+	registerAction('fields.create', ({ key, payload = {} }) => ({
+		collection: 'directus_fields',
+		action: 'create',
+		key,
+		payload,
+	}));
+	registerAction('fields.update', ({ keys, payload = {} }) => ({
+		collection: 'directus_fields',
+		action: 'update',
+		keys,
+		payload,
+	}));
+	registerAction('fields.delete', ({ keys, payload = [] }) => ({
+		collection: 'directus_fields',
+		action: 'delete',
+		keys,
+		payload,
+	}));
+}
+
+function registerFilesHooks() {
+	// extra event for file uploads that doubles as create event
+	registerAction('files.upload', ({ key, collection, payload = {} }) => ({
+		collection,
+		action: 'create',
+		key,
+		payload,
+	}));
+	registerAction('files.update', ({ keys, collection, payload = {} }) => ({
+		collection,
+		action: 'update',
+		keys,
+		payload,
+	}));
+	registerAction('files.delete', ({ keys, collection, payload = [] }) => ({
+		collection,
+		action: 'delete',
+		keys,
+		payload,
+	}));
+}
+
+function registerRelationsHooks() {
+	// exception for relation updates that don't report `directus_relations` as being the collection
```

Comment:
```suggestion
	// exception for relation hooks that don't report `directus_relations` as being the collection
```

(just to disambiguate it with `update` event specifically)